ssz_serialization/merkleization.nim

```diff
@@ -0,0 +1,734 @@
+# ssz_serialization
+# Copyright (c) 2018-2021 Status Research & Development GmbH
+# Licensed and distributed under either of
+#   * MIT license (license terms in the root directory or at https://opensource.org/licenses/MIT).
+#   * Apache v2 license (license terms in the root directory or at https://www.apache.org/licenses/LICENSE-2.0).
+# at your option. This file may not be copied, modified, or distributed except according to those terms.
+
+# This module contains the parts necessary to create a merkle hash from the core
+# SSZ types outlined in the spec:
+# https://github.com/ethereum/consensus-specs/blob/v1.0.1/ssz/simple-serialize.md#merkleization
+
+{.push raises: [Defect].}
+
+import
+  stew/[bitops2, endians2, ptrops],
+  stew/ranges/ptr_arith, nimcrypto/[hash, sha2],
+  serialization/testing/tracing,
+  "."/[bitseqs, codec, types]
+
+const PREFER_BLST_SHA256* {.booldefine.} = true
+
+when PREFER_BLST_SHA256:
+  import blscurve
+  when BLS_BACKEND == BLST:
+    const USE_BLST_SHA256 = true
+  else:
+    const USE_BLST_SHA256 = false
+else:
+  const USE_BLST_SHA256 = false
+
+export
+  codec, bitseqs, types
+
+when hasSerializationTracing:
+  import stew/byteutils, typetraits
+
+const
+  zero64 = default array[64, byte]
+  bitsPerChunk = bytesPerChunk * 8
+
+func binaryTreeHeight*(totalElements: Limit): int =
+  bitWidth nextPow2(uint64 totalElements)
+
+type
+  # TODO Figure out what would be the right type for this.
+  #      It probably fits in uint16 for all practical purposes.
+  GeneralizedIndex* = uint32
+
+  SszMerkleizerImpl = object
+    combinedChunks: ptr UncheckedArray[Digest]
+    totalChunks: uint64
+    topIndex: int
+
+  SszMerkleizer*[limit: static[Limit]] = object
+    combinedChunks: ref array[binaryTreeHeight limit, Digest]
+    impl: SszMerkleizerImpl
+
+template chunks*(m: SszMerkleizerImpl): openArray[Digest] =
+  m.combinedChunks.toOpenArray(0, m.topIndex)
+
+template getChunkCount*(m: SszMerkleizer): uint64 =
+  m.impl.totalChunks
+
+template getCombinedChunks*(m: SszMerkleizer): openArray[Digest] =
+  toOpenArray(m.impl.combinedChunks, 0, m.impl.topIndex)
+
+when USE_BLST_SHA256:
+  export blscurve.update
+  type DigestCtx* = BLST_SHA256_CTX
+else:
+  type DigestCtx* = sha2.sha256
+
+template computeDigest*(body: untyped): Digest =
```

Comment:
`Eth2Digest` -> `Digest` and the move of this template (and its when clauses) is basically the only "real" change in this PR.